File: bench/cross-language/tree/tree.go
package main

import "fmt"

type Node struct {
	left  *Node
	right *Node
	value int
}

func buildTree(depth int) *Node {
	if depth <= 0 {
		return &Node{value: 1}
	}
	return &Node{
		left:  buildTree(depth - 1),
		right: buildTree(depth - 1),
	}
}

func checkTree(node *Node) int {
	if node.left == nil {
		return node.value
	}
	return checkTree(node.left) + checkTree(node.right)
}

func main() {
	tree := buildTree(18)
	checksum := checkTree(tree)
	fmt.Printf("Checksum: %d\n", checksum)
}

File: bench/cross-language/tree/tree.py
import sys

sys.setrecursionlimit(1000000)


class Node:
    __slots__ = ["left", "right", "value"]

    def __init__(self, left=None, right=None, value=0):
        self.left = left
        self.right = right
        self.value = value


def build_tree(depth: int) -> Node:
    if depth <= 0:
        return Node(value=1)
    return Node(
        left=build_tree(depth - 1),
        right=build_tree(depth - 1),
    )


def check_tree(node: Node) -> int:
    if node.left is None:
        return node.value
    return check_tree(node.left) + check_tree(node.right)


if __name__ == "__main__":
    tree = build_tree(18)
    checksum = check_tree(tree)
    print(f"Checksum: {checksum}")
